Extended claim types

rasmusguterstam opened the pull request:
# Jira Issue: [ [MX-143](https://hedvig.atlassian.net/secure/RapidBoard.jspa?rapidView=12&projectKey=MX&modal=detail&selectedIssue=MX-143) & [MX-145](https://hedvig.atlassian.net/secure/RapidBoard.jspa?rapidView=12&projectKey=MX&modal=detail&selectedIssue=MX-145) ]

Associated with [this back-office PR](https://github.com/HedvigInsurance/back-office/pull/360)

## What?
- Implement a new claim type model which goes from introduces: `ClaimType`, `ClaimSubType` and `ClaimFlag`
- Implement an upcaster for `ClaimsTypeUpdateEvent`
- Implement a `ClaimTypeMapper`, which maps both legacy claim types (from events) and hedgehog claim types (😋 🔫)
 
## Why?
- Claim types are too general or not general enough
- Certain claim types used right now are not really claim *types* per se (`Not Covered,` `Test`, `Duplicate Claim`, `Confirmed Fraud`). Moreover, they're not necessarily mutually exclusive.
- We want to add `Accident` to claim types, which is tedious and difficult to do with current setup
- Claim types are currently spread out across `hedgehog`, `back-office` and `claims-service`, and this is the first step in moving the source of truth to `claims-service`

## Misc
- This implementation makes it possible to integrate the new claim type model internally in `claims-service` while at the same time leaving the interface with respect to `back-office` intact. This way we can separate concerns quite nicely 🤝
- Elvin and I discussed whether we should create a specific event for updating `ClaimFlag`, but this seemed like the easiest step as of the current implementation (instead of upcasting 1 event to 2, which seems weird?). Any thoughts on this?
- Naming might have escalated a bit, would love some feedback
- And last but not least, would love some feedback on the modelling 😇 (`ClaimType`, `ClaimSubType`, `ClaimFlag`)

## Optional checklist
- [x] Codescouted
- [x] Unit tests written
- [x] Tested locally

bystam commented on src/main/java/com/hedvig/claims/commands/UpdateClaimTypeCommand.kt:
```diff
@@ -0,0 +1,51 @@
+        this.flag = flag?.name
+        log.info(this.toString())
+    }
+}
```
To me it looks like this could/should be a data class, wish all the `var` properties being `val` and in the constructor? And then you could drop the amount of code in this file down to like one third.

Could it be possible to also either avoid the second `log.info` in here, or put it inside the first log statement? A simple toStringed version of an object in the logs does not provide very much context.

bystam commented on src/main/java/com/hedvig/claims/events/upcast/ClaimsTypeUpdateEvent_v1.kt:
```diff
@@ -0,0 +1,55 @@
+        )
+
+        private val outputType = SimpleSerializedType(
+            ClaimsTypeUpdateEvent::class.java.typeName, "1.0"
```
I propose not importing `ClaimsTypeUpdateEvent` at all here and instead spelling out its name including the package as a hard coded string. If `ClaimsTypeUpdateEvent` were to be moved in the codebase, this migration will break as it actually should target the historical, persisted form of these types. That's why the `SimpleSerializedType` object takes a strings and not a `Class` in the first place.

So I think both of them should contain `"com.hedvig.claims.events.ClaimsTypeUpdateEvent"`

rasmusguterstam replied:
Never thought about it that way - that's a really good point!

bystam commented on src/main/java/com/hedvig/claims/query/ClaimReportHistoryEventListener.java:
```diff
@@ -104,6 +104,7 @@ public void on(ClaimsTypeUpdateEvent e, @Timestamp Instant timestamp) {
 
     ClaimReportHistoryEntity claimHistoryEntry = copyLatestClaimHistoryEntity(e.claimID, timestamp);
 
+    // TODO: Don't wanna mess up reporting
```
is this a todo for this PR or a future one?

rasmusguterstam replied:
It has been cleared with Akash now, so we're good 🤝

bystam commented on src/test/java/com/hedvig/claims/upcast/ClaimsTypeUpdateEventV1Test.kt:
```diff
@@ -0,0 +1,99 @@
+        val upcastedEvent =
+            serializer.deserialize<ClaimsTypeUpdateEvent, ClaimsTypeUpdateEvent>(event.data as SerializedObject<ClaimsTypeUpdateEvent>)
+
+        println(upcastedEvent)
```
Was this used for debugging the test? If so I think we can omit it now

rasmusguterstam replied:
🤦‍♂️

bystam reviewed (commented):
This looks nice! :) I have some idea for it though, mostly small

bystam commented on src/main/java/com/hedvig/claims/aggregates/ClaimsAggregate.java:
```diff
@@ -81,7 +81,11 @@
+
+    public ClaimType type;
+    public ClaimSubType subType;
+    public ClaimFlag flag;
```
Or keep the enum but make it a list/set of them? Personally I think that's the most common way to do it.

rasmusguterstam commented on src/main/java/com/hedvig/claims/web/dto/ClaimTypeDTO.kt:
```diff
@@ -0,0 +1,7 @@
+    var type: String? = null
+    var subType: String? = null
+    var flag: String? = null
+}
```
Yeah I agree, always makes you look up the original DTO either way 😅

rasmusguterstam commented on src/main/java/com/hedvig/claims/aggregates/ClaimType.kt:
```diff
@@ -0,0 +1,90 @@
+    ACCIDENT(displayName = "Accident");
+
+    companion object {
+        fun getSubTypes(type: ClaimType): List<ClaimSubType> = ClaimSubType.values().filter { it.parentType == type }
```
That's really neat! And yes, it is unused for now, but it was something I did in the beginning while trying to decide in which PR I will create the end-point for back-office & hedgehog to retrieve the types (which turned out to not be this one, but it's incoming)

rasmusguterstam commented:
## Some changes since previous review:

- Decopule `ClaimType` and `ClaimSubType`, but add a `Set` of `ClaimSubTypes` to the `ClaimType` enum. We could simply go for a list of strings here, but considering the very high probability of someone changing their mind about a specific name it seems reasonable that we still have the possibility to change `displayName` while keeping the model intact.
- Go from `SingleEventUpcaster` to `EventMultiUpcaster` for the upcasting of `ClaimsTypeUpdateEvent`. In line with this, we also introduce distinct events for the new fields: `ClaimSubTypeUpdateEvent`, `ClaimFlagSetEvent` and `ClaimFlagsResetEvent` (which clears all current `ClaimFlags` for a claim)

bystam commented on src/main/java/com/hedvig/claims/commands/UpdateClaimSubTypeCommand.kt:
```diff
@@ -0,0 +1,12 @@
+    val claimsId: String,
+    val userId: String?,
+    val subType: ClaimSubType?
+    )
```
whitespace is off

bystam commented on src/main/java/com/hedvig/claims/commands/SetClaimFlagCommand.kt:
```diff
@@ -0,0 +1,12 @@
+    val claimId: String,
+    val userId: String?,
+    val flag: ClaimFlag,
+    val flagValue: Boolean,
```
Is `flagValue` whether it should be enabled or not? Maybe `enabled` or `isEnabled` is more readable in that case? :)

rasmusguterstam replied:
Yes indeed!

bystam commented on src/main/java/com/hedvig/claims/query/ClaimEntity.java:
```diff
@@ -33,6 +27,15 @@
+  public ClaimSubType subType;
+
+  @ElementCollection
+  @JoinColumn(name = "claimsId")
```
If you're 100% sure this is correct then go ahead :) it's just that my understanding of what `@JoinColumn` is does not add up with this usage really. But as I said I am not great at all this.

rasmusguterstam replied:
Haha I'm not sure it's the correct one at all. Tried to google as well, and as we're doing this on other columns on `ClaimEntity` (not necessarily sets of enums though) I just went with it. But good catch, database-wise it seems to look 
 and work the same, but we don't want any weird unintendended consequences later on - this def looks better!

bystam reviewed (commented):
Major improvements from the last version, I like it! Just had a couple of questions. One of them a bit bigger perhaps.

Michael-Duivestein commented on src/main/java/com/hedvig/claims/aggregates/ClaimsAggregate.java:
```diff
@@ -3,56 +3,16 @@
-import com.hedvig.claims.events.IndemnityCostPaymentAddedEvent;
-import com.hedvig.claims.events.NoteAddedEvent;
-import com.hedvig.claims.events.PaymentAddedEvent;
+import com.hedvig.claims.commands.*;
+import com.hedvig.claims.events.*;
```
We're trying to get rid of * imports.

Michael-Duivestein commented on src/main/java/com/hedvig/claims/aggregates/ClaimsAggregate.java:
```diff
@@ -81,7 +41,11 @@
+
+    public ClaimType type;
+    public ClaimSubType subType;
+    public Set<ClaimFlag> flags;
```
Nice!

Michael-Duivestein reviewed (commented):
I really like what you did with the claim types, subtypes, and flags.

bystam reviewed (approved):
Great job!

vonElfvin commented:
With the current question marks regarding what claim types Hedvig wants we'll convert this to a draft until later. The code has untangled a lot of problems we need to address however.
